gilbreth_perception/src/recognition_node.cpp

```diff
@@ -0,0 +1,295 @@
+
+	// Load model settings
+	XmlRpc::XmlRpcValue model_map;
+	std::map<std::string,float> pick_orientation;
+	std::string package_path;
+	ros::NodeHandle ph("~");
+	ph.getParam("part_list",model_map);
+	ph.getParam("pick_orientation",pick_orientation);
+	ph.getParam("package_path",package_path);
+
+	std::vector<int> pick_point_indices;
+	pcl::PointCloud<PointType>::Ptr scene (new pcl::PointCloud<PointType> ());
+	pcl::PointCloud<NormalType>::Ptr scene_normals (new pcl::PointCloud<NormalType> ());
+	pcl::PointCloud<pcl::FPFHSignature33>::Ptr scene_features (new pcl::PointCloud<pcl::FPFHSignature33> );
+	pcl::PassThrough<PointType> pass;
+	pcl::SACSegmentationFromNormals<PointType, pcl::Normal> seg;
+	pcl::search::KdTree<PointType>::Ptr tree (new pcl::search::KdTree<PointType> ());
+	std::vector<pcl::PointCloud<PointType>::Ptr> model_raw_list;
+	std::vector<std::string> model_name;
+	for(int i=0;i<model_map.size();i++){
+		pcl::PointCloud<PointType>::Ptr model_raw (new pcl::PointCloud<PointType> ());
+		std::string model_path=model_map[i]["path"];
+
+		if (pcl::io::loadPCDFile (package_path + model_path, *model_raw) < 0)
+		{
+			std::cout << "Error loading model cloud." << std::endl;
+			return ;
+		}
+		model_raw_list.push_back (model_raw);
+		model_name.push_back(model_map[i]["name"]);
+		pick_point_indices.push_back(model_map[i]["pick_point_index"]);
+
+	}
+
+	//Load scene
+	pcl::fromROSMsg(*cloud_msg, *scene);
+
+	//Downsample models
+	pcl::VoxelGrid<pcl::PointXYZ> sor;
+	std::vector<pcl::PointCloud<PointType>::Ptr> model_list;
+	for (int i=0; i< model_raw_list.size(); i++){
+
+		pcl::PointCloud<PointType>::Ptr model (new pcl::PointCloud<PointType> ());
+		sor.setInputCloud (model_raw_list[i]);
+		sor.setLeafSize (down_sample_, down_sample_, down_sample_);
+		sor.filter (*model);
+		model_list.push_back(model);
+
+	}
+
+
+	//  Compute Feature
+	pcl::NormalEstimationOMP<PointType, NormalType> norm_est;
+	std::vector<pcl::PointCloud<NormalType>::Ptr> model_normals_list;
+	pcl::FPFHEstimation<pcl::PointXYZ, pcl::Normal, pcl::FPFHSignature33> fpfh_est;
+	std::vector<pcl::PointCloud<pcl::FPFHSignature33>::Ptr> model_features_list;
+
+	norm_est.setSearchMethod (tree);
+	norm_est.setRadiusSearch (descr_rad_);
+	norm_est.setInputCloud (scene);
+	norm_est.compute (*scene_normals);
+
+	fpfh_est.setSearchMethod (tree);
+	fpfh_est.setRadiusSearch (descr_rad_);
+	fpfh_est.setInputCloud (scene);
+	fpfh_est.setInputNormals (scene_normals);
+	fpfh_est.compute (*scene_features);
+
+	for(int i=0;i<model_list.size();i++){
+		pcl::PointCloud<NormalType>::Ptr model_normals (new pcl::PointCloud<NormalType> ());
+		norm_est.setRadiusSearch (descr_rad_);
+		norm_est.setSearchMethod(tree);
+		norm_est.setInputCloud (model_list[i]);
+		norm_est.compute (*model_normals);
+		model_normals_list.push_back(model_normals);
+	}
+
+	for(int i=0;i<model_list.size();i++){
+		pcl::PointCloud<pcl::FPFHSignature33>::Ptr model_features (new pcl::PointCloud<pcl::FPFHSignature33> );
```

Comment:
Use two space indentation.  You can run a [formatter](http://moveit.ros.org/documentation/contributing/code/) to automatically fix indentation and other formatting conventions